rust/protocol/src/error.rs

```diff
@@ -11,82 +15,208 @@ use uuid::Uuid;
+    ///
+    /// Prefer to raise [Self::InvalidState] except in methods which directly decode protobufs.
     InvalidProtobufEncoding,
 
     /// ciphertext serialized bytes were too short <{0}>
+    ///
+    /// Raised if some ciphertext was deserialized from a too-small slice.
+    ///
+    /// Prefer to make API method signatures and wrapper structs consume and produce static-sized
+    /// byte slices to minimize the need to raise this error.
     CiphertextMessageTooShort(usize),
     /// ciphertext version was too old <{0}>
+    ///
+    /// Raised if the ciphertext version decoded from a protobuf is older than this client.
+    ///
+    /// The current client's ciphertext version is at [CIPHERTEXT_MESSAGE_CURRENT_VERSION].
     LegacyCiphertextVersion(u8),
     /// ciphertext version was unrecognized <{0}>
+    ///
+    /// Raised if the ciphertext version decoded from a protobuf is newer than this client.
+    ///
+    /// The current client's ciphertext version is at [CIPHERTEXT_MESSAGE_CURRENT_VERSION].
     UnrecognizedCiphertextVersion(u8),
     /// unrecognized message version <{0}>
+    ///
+    /// Raised if the ciphertext version decoded from a protobuf fails to match the cached version
+    /// for the message chain that message originates from.
+    ///
+    /// *TODO: This case should wrap the same numeric type as [Self::LegacyCiphertextVersion] and
+    /// [Self::UnrecognizedCiphertextVersion]. This dissonance is addressed in
+    /// <https://github.com/signalapp/libsignal-client/pull/289>.*
     UnrecognizedMessageVersion(u32),
 
     /// fingerprint version number mismatch them {0} us {1}
+    ///
+    /// Raised if a fingerprint version decoded from a protobuf has an unexpected value.
     FingerprintVersionMismatch(u32, u32),
     /// fingerprint parsing error
+    ///
+    /// Raised if a field in a fingerprint protobuf is invalid in some way.
+    ///
+    /// Similar to [Self::InvalidProtobufEncoding].
     FingerprintParsingError,
 
     /// no key type identifier
+    ///
+    /// Raised if a [PublicKey] is deserialized from an empty slice.
+    ///
+    /// Prefer to use static-sized slices in API method signatures and struct fields to minimize the
+    /// need to raise this error.
     NoKeyTypeIdentifier,
     /// bad key type <{0:#04x}>
+    ///
+    /// Raised if a [KeyType] decoded from a [u8] has an unrecognized value.
     BadKeyType(u8),
     /// bad key length <{1}> for key with type <{0}>
+    ///
+    /// Raised if a [PublicKey] or [PrivateKey] is deserialized from a slice of incorrect length.
+    ///
+    /// Prefer to use static-sized slices in API method signatures and struct fields to minimize the
+    /// need to raise this error.
     BadKeyLength(KeyType, usize),
 
     /// invalid signature detected
+    ///
+    /// Raised if signature validation fails for a [SignedPreKeyRecord] or a [SenderKeyMessage].
     SignatureValidationFailed,
```

Comment:
Nitpick: it would be better to make it clear that these are examples; many things have signatures that could fail to validate. These are just the two we surface as "signature validation failed".